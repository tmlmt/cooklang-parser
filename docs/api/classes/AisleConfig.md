[cooklang-parser](../globals.md) / AisleConfig

# Class: AisleConfig

Defined in: classes/aisle\_config.ts:7

Represents the aisle configuration for a shopping list.

## Constructors

### Constructor

> **new AisleConfig**(`config?`): `AisleConfig`

Defined in: classes/aisle\_config.ts:18

Creates a new AisleConfig instance.

#### Parameters

##### config?

`string`

The aisle configuration to parse.

#### Returns

`AisleConfig`

## Properties

### categories

> **categories**: [`AisleCategory`](../interfaces/AisleCategory.md)[] = `[]`

Defined in: classes/aisle\_config.ts:12

The categories of aisles.

#### See

[AisleCategory](../interfaces/AisleCategory.md)

## Methods

### parse()

> **parse**(`config`): `void`

Defined in: classes/aisle\_config.ts:28

Parses an aisle configuration from a string.

#### Parameters

##### config

`string`

The aisle configuration to parse.

#### Returns

`void`
