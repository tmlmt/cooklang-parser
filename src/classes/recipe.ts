import type {
  Metadata,
  Ingredient,
  IngredientItem,
  Timer,
  Step,
  Note,
  Cookware,
  MetadataExtract,
  CookwareItem,
} from "../types";
import { Section } from "./section";
import {
  tokensRegex,
  commentRegex,
  blockCommentRegex,
  metadataRegex,
  ingredientAliasRegex,
  floatRegex,
} from "../regex";
import {
  flushPendingItems,
  flushPendingNote,
  findAndUpsertIngredient,
  findAndUpsertCookware,
  parseQuantityInput,
  extractMetadata,
} from "../parser_helpers";
import {
  addQuantities,
  getDefaultQuantityValue,
  multiplyQuantityValue,
  type Quantity,
} from "../units";

/**
 * Recipe parser.
 *
 * ## Usage
 *
 * You can either directly provide the recipe string when creating the instance
 * e.g. `const recipe = new Recipe('Add @eggs{3}')`, or create it first and then pass
 * the recipe string to the {@link Recipe.parse | parse()} method.
 *
 * Look at the [properties](#properties) to see how the recipe's properties are parsed.
 *
 * @category Classes
 *
 * @example
 * ```typescript
 * import { Recipe } from "@tmlmt/cooklang-parser";
 *
 * const recipeString = `
 * ---
 * title: Pancakes
 * tags: [breakfast, easy]
 * ---
 * Crack the @eggs{3} with @flour{100%g} and @milk{200%mL}
 *
 * Melt some @butter{50%g} in a #pan on medium heat.
 *
 * Cook for ~{5%minutes} on each side.
 * `
 * const recipe = new Recipe(recipeString);
 * ```
 */
export class Recipe {
  /**
   * The parsed recipe metadata.
   */
  metadata: Metadata = {};
  /**
   * The parsed recipe ingredients.
   */
  ingredients: Ingredient[] = [];
  /**
   * The parsed recipe sections.
   */
  sections: Section[] = [];
  /**
   * The parsed recipe cookware.
   */
  cookware: Cookware[] = [];
  /**
   * The parsed recipe timers.
   */
  timers: Timer[] = [];
  /**
   * The parsed recipe servings. Used for scaling. Parsed from one of
   * {@link Metadata.servings}, {@link Metadata.yield} or {@link Metadata.serves}
   * metadata fields.
   *
   * @see {@link Recipe.scaleBy | scaleBy()} and {@link Recipe.scaleTo | scaleTo()} methods
   */
  servings?: number;

  /**
   * Creates a new Recipe instance.
   * @param content - The recipe content to parse.
   */
  constructor(content?: string) {
    if (content) {
      this.parse(content);
    }
  }

  /**
   * Parses a recipe from a string.
   * @param content - The recipe content to parse.
   */
  parse(content: string) {
    const cleanContent = content
      .replace(metadataRegex, "")
      .replace(commentRegex, "")
      .replace(blockCommentRegex, "")
      .trim()
      .split(/\r\n?|\n/);

    const { metadata, servings }: MetadataExtract = extractMetadata(content);
    this.metadata = metadata;
    this.servings = servings;

    let blankLineBefore = true;
    let section: Section = new Section();
    const items: Step["items"] = [];
    let note: Note["note"] = "";
    let inNote = false;

    for (const line of cleanContent) {
      if (line.trim().length === 0) {
        flushPendingItems(section, items);
        note = flushPendingNote(section, note);
        blankLineBefore = true;
        inNote = false;
        continue;
      }

      if (line.startsWith("=")) {
        flushPendingItems(section, items);
        note = flushPendingNote(section, note);

        if (this.sections.length === 0 && section.isBlank()) {
          section.name = line.substring(1).trim();
        } else {
          if (!section.isBlank()) {
            this.sections.push(section);
          }
          section = new Section(line.substring(1).trim());
        }
        blankLineBefore = true;
        inNote = false;
        continue;
      }

      if (blankLineBefore && line.startsWith(">")) {
        flushPendingItems(section, items);
        note = flushPendingNote(section, note);
        note += line.substring(1).trim();
        inNote = true;
        blankLineBefore = false;
        continue;
      }

      if (inNote) {
        if (line.startsWith(">")) {
          note += " " + line.substring(1).trim();
        } else {
          note += " " + line.trim();
        }
        blankLineBefore = false;
        continue;
      }

      note = flushPendingNote(section, note);

      let cursor = 0;
      for (const match of line.matchAll(tokensRegex)) {
        const idx = match.index;
        if (idx > cursor) {
          items.push({ type: "text", value: line.slice(cursor, idx) });
        }

        const groups = match.groups!;

        if (groups.mIngredientName || groups.sIngredientName) {
          const name = (groups.mIngredientName || groups.sIngredientName)!;
          const scalableQuantity =
            (groups.mIngredientQuantityModifier ||
              groups.sIngredientQuantityModifier) !== "=";
          const quantityRaw =
            groups.mIngredientQuantity || groups.sIngredientQuantity;
          const unit = groups.mIngredientUnit || groups.sIngredientUnit;
          const preparation =
            groups.mIngredientPreparation || groups.sIngredientPreparation;
          const modifier =
            groups.mIngredientModifier || groups.sIngredientModifier;
          const optional = modifier === "?";
          const hidden = modifier === "-";
          const reference = modifier === "&";
          const isRecipe = modifier === "@";
          const quantity = quantityRaw
            ? parseQuantityInput(quantityRaw)
            : undefined;
          const aliasMatch = name.match(ingredientAliasRegex);
          let listName, displayName: string;
          if (
            aliasMatch &&
            aliasMatch.groups!.ingredientListName!.trim().length > 0 &&
            aliasMatch.groups!.ingredientDisplayName!.trim().length > 0
          ) {
            listName = aliasMatch.groups!.ingredientListName!.trim();
            displayName = aliasMatch.groups!.ingredientDisplayName!.trim();
          } else {
            listName = name;
            displayName = name;
          }

          const idxsInList = findAndUpsertIngredient(
            this.ingredients,
            {
              name: listName,
              quantity,
              quantityParts: quantity
                ? [
                    {
                      value: quantity,
                      unit,
                      scalable: scalableQuantity,
                    },
                  ]
                : undefined,
              unit,
              optional,
              hidden,
              preparation,
              isRecipe,
            },
            reference,
          );

          const newItem: IngredientItem = {
            type: "ingredient",
            index: idxsInList.ingredientIndex,
            displayName,
          };
          if (idxsInList.quantityPartIndex !== undefined) {
            newItem.quantityPartIndex = idxsInList.quantityPartIndex;
          }
          items.push(newItem);
        } else if (groups.mCookwareName || groups.sCookwareName) {
          const name = (groups.mCookwareName || groups.sCookwareName)!;
          const modifier = groups.mCookwareModifier || groups.sCookwareModifier;
          const quantityRaw =
            groups.mCookwareQuantity || groups.sCookwareQuantity;
          const optional = modifier === "?";
          const hidden = modifier === "-";
          const reference = modifier === "&";
          const quantity = quantityRaw
            ? parseQuantityInput(quantityRaw)
            : undefined;

          const idxsInList = findAndUpsertCookware(
            this.cookware,
            {
              name,
              quantity,
              quantityParts: quantity ? [quantity] : undefined,
              optional,
              hidden,
            },
            reference,
          );
          items.push({
            type: "cookware",
            index: idxsInList.cookwareIndex,
            quantityPartIndex: idxsInList.quantityPartIndex,
          } as CookwareItem);
        } else if (groups.timerQuantity !== undefined) {
          const durationStr = groups.timerQuantity.trim();
          const unit = (groups.timerUnit || "").trim();
          if (!unit) {
            throw new Error("Timer missing unit");
          }
          const name = groups.timerName || undefined;
          const duration = parseQuantityInput(durationStr);
          const timerObj: Timer = {
            name,
            duration,
            unit,
          };
          items.push({ type: "timer", index: this.timers.push(timerObj) - 1 });
        }

        cursor = idx + match[0].length;
      }

      if (cursor < line.length) {
        items.push({ type: "text", value: line.slice(cursor) });
      }

      blankLineBefore = false;
    }

    // End of content reached: pushing all temporarily saved elements
    flushPendingItems(section, items);
    note = flushPendingNote(section, note);
    if (!section.isBlank()) {
      this.sections.push(section);
    }
  }

  /**
   * Scales the recipe to a new number of servings. In practice, it calls
   * {@link Recipe.scaleBy | scaleBy} with a factor corresponding to the ratio between `newServings`
   *   and the recipe's {@link Recipe.servings | servings} value.
   * @param newServings - The new number of servings.
   * @returns A new Recipe instance with the scaled ingredients.
   * @throws `Error` if the recipe does not contains an initial {@link Recipe.servings | servings} value
   */
  scaleTo(newServings: number): Recipe {
    const originalServings = this.getServings();

    if (originalServings === undefined || originalServings === 0) {
      throw new Error("Error scaling recipe: no initial servings value set");
    }

    const factor = newServings / originalServings;
    return this.scaleBy(factor);
  }

  /**
   * Scales the recipe by a factor.
   * @param factor - The factor to scale the recipe by.
   * @returns A new Recipe instance with the scaled ingredients.
   */
  scaleBy(factor: number): Recipe {
    const newRecipe = this.clone();

    const originalServings = newRecipe.getServings();

    if (originalServings === undefined || originalServings === 0) {
      throw new Error("Error scaling recipe: no initial servings value set");
    }

    newRecipe.ingredients = newRecipe.ingredients
      .map((ingredient) => {
        // Scale first individual parts of total quantity depending on whether they are scalable or not
        if (ingredient.quantityParts) {
          ingredient.quantityParts = ingredient.quantityParts.map(
            (quantityPart) => {
              if (
                quantityPart.value.type === "fixed" &&
                quantityPart.value.value.type === "text"
              ) {
                return quantityPart;
              }
              return {
                ...quantityPart,
                value: multiplyQuantityValue(
                  quantityPart.value,
                  quantityPart.scalable ? factor : 1,
                ),
              };
            },
          );
          // Recalculate total quantity from quantity parts
          if (ingredient.quantityParts.length === 1) {
            ingredient.quantity = ingredient.quantityParts[0]!.value;
            ingredient.unit = ingredient.quantityParts[0]!.unit;
          } else {
            const totalQuantity = ingredient.quantityParts.reduce(
              (acc, val) =>
                addQuantities(acc, { value: val.value, unit: val.unit }),
              { value: getDefaultQuantityValue() } as Quantity,
            );
            ingredient.quantity = totalQuantity.value;
            ingredient.unit = totalQuantity.unit;
          }
        }
        return ingredient;
      })
      .filter((ingredient) => ingredient.quantity !== null);

    newRecipe.servings = originalServings * factor;

    if (newRecipe.metadata.servings && this.metadata.servings) {
      if (
        floatRegex.test(String(this.metadata.servings).replace(",", ".").trim())
      ) {
        const servingsValue = parseFloat(
          String(this.metadata.servings).replace(",", "."),
        );
        newRecipe.metadata.servings = String(servingsValue * factor);
      }
    }

    if (newRecipe.metadata.yield && this.metadata.yield) {
      if (
        floatRegex.test(String(this.metadata.yield).replace(",", ".").trim())
      ) {
        const yieldValue = parseFloat(
          String(this.metadata.yield).replace(",", "."),
        );
        newRecipe.metadata.yield = String(yieldValue * factor);
      }
    }

    if (newRecipe.metadata.serves && this.metadata.serves) {
      if (
        floatRegex.test(String(this.metadata.serves).replace(",", ".").trim())
      ) {
        const servesValue = parseFloat(
          String(this.metadata.serves).replace(",", "."),
        );
        newRecipe.metadata.serves = String(servesValue * factor);
      }
    }

    return newRecipe;
  }

  /**
   * Gets the number of servings for the recipe.
   * @private
   * @returns The number of servings, or undefined if not set.
   */
  private getServings(): number | undefined {
    if (this.servings) {
      return this.servings;
    }
    return undefined;
  }

  /**
   * Clones the recipe.
   * @returns A new Recipe instance with the same properties.
   */
  clone(): Recipe {
    const newRecipe = new Recipe();
    // deep copy
    newRecipe.metadata = JSON.parse(JSON.stringify(this.metadata)) as Metadata;
    newRecipe.ingredients = JSON.parse(
      JSON.stringify(this.ingredients),
    ) as Ingredient[];
    newRecipe.sections = JSON.parse(JSON.stringify(this.sections)) as Section[];
    newRecipe.cookware = JSON.parse(
      JSON.stringify(this.cookware),
    ) as Cookware[];
    newRecipe.timers = JSON.parse(JSON.stringify(this.timers)) as Timer[];
    newRecipe.servings = this.servings;
    return newRecipe;
  }
}
