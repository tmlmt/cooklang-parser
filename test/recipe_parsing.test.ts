import { describe, it, expect } from "vitest";
import { Recipe } from "../src/classes/recipe";
import { simpleRecipe, complexRecipe } from "./fixtures/recipes";

describe("parse function", () => {
  it("parses basic metadata correctly", () => {
    const result = new Recipe(simpleRecipe);
    expect(result.metadata.title).toBe("Pancakes");
    expect(result.metadata.tags).toEqual(["breakfast", "easy"]);
  });

  it("extracts ingredients correctly", () => {
    const result = new Recipe(simpleRecipe);
    expect(result.ingredients.length).toBe(4);
    expect(result.ingredients).toEqual([
      {
        name: "eggs",
        optional: false,
        hidden: false,
        quantity: {
          type: "fixed",
          value: { type: "decimal", value: 3 },
        },
        unit: undefined,
        preparation: undefined,
        isRecipe: false,
      },
      {
        name: "flour",
        optional: false,
        hidden: false,
        quantity: undefined,
        unit: undefined,
        preparation: undefined,
        isRecipe: false,
      },
      {
        name: "coarse salt",
        hidden: false,
        isRecipe: false,
        optional: false,
        preparation: undefined,
        quantity: undefined,
        unit: undefined,
      },
      {
        name: "butter",
        optional: false,
        hidden: false,
        quantity: {
          type: "fixed",
          value: { type: "decimal", value: 50 },
        },
        unit: "g",
        preparation: undefined,
        isRecipe: false,
      },
    ]);
  });

  it("parses ingredients that are other recipes", () => {
    const recipe = `
      Defrost @@pizza dough{1} and form it into a nice disc
      And @@toppings on top
    `;
    const result = new Recipe(recipe);
    expect(result.ingredients).toHaveLength(2);
    expect(result.ingredients[0]).toEqual({
      name: "pizza dough",
      quantity: { type: "fixed", value: { type: "decimal", value: 1 } },
      unit: undefined,
      preparation: undefined,
      optional: false,
      hidden: false,
      isRecipe: true,
    });
    expect(result.ingredients[1]).toEqual({
      name: "toppings",
      quantity: undefined,
      unit: undefined,
      preparation: undefined,
      optional: false,
      hidden: false,
      isRecipe: true,
    });
  });

  it("parses ingredients with preparation", () => {
    const recipe = `
      Add some @wheat flour{100%g}(sifted).
      And @eggs{2}(large, beaten).
    `;
    const result = new Recipe(recipe);
    expect(result.ingredients).toHaveLength(2);
    expect(result.ingredients[0]).toEqual({
      name: "wheat flour",
      quantity: { type: "fixed", value: { type: "decimal", value: 100 } },
      unit: "g",
      preparation: "sifted",
      optional: false,
      hidden: false,
      isRecipe: false,
    });
    expect(result.ingredients[1]).toEqual({
      name: "eggs",
      quantity: { type: "fixed", value: { type: "decimal", value: 2 } },
      unit: undefined,
      preparation: "large, beaten",
      optional: false,
      hidden: false,
      isRecipe: false,
    });
  });

  it("parses hidden and optional ingredients", () => {
    const recipe = `
      Add some @-salt{}.
      And maybe some @?pepper{}.
    `;
    const result = new Recipe(recipe);
    expect(result.ingredients).toHaveLength(2);
    expect(result.ingredients[0]).toEqual({
      name: "salt",
      hidden: true,
      optional: false,
      quantity: undefined,
      unit: undefined,
      preparation: undefined,
      isRecipe: false,
    });
    expect(result.ingredients[1]).toEqual({
      name: "pepper",
      hidden: false,
      optional: true,
      quantity: undefined,
      unit: undefined,
      preparation: undefined,
      isRecipe: false,
    });
  });

  describe("ingredient referencing", () => {
    it("should add quantities of explicitly referenced ingredients with same units", () => {
      const recipe = `
        Add @flour{100%g}.
        Then add some more @&flour{50%g}.
      `;
      const result = new Recipe(recipe);
      expect(result.ingredients).toHaveLength(1);
      expect(result.ingredients[0]).toEqual({
        name: "flour",
        quantity: { type: "fixed", value: { type: "decimal", value: 150 } },
        unit: "g",
        optional: false,
        hidden: false,
        preparation: undefined,
        isRecipe: false,
      });
    });

    it("should list ingredients separately if not referenced with '&'", () => {
      const recipe = `
        Add @flour{100%g}.
        Then add some more @flour{50%g}.
      `;
      const result = new Recipe(recipe);
      expect(result.ingredients).toHaveLength(2);
      expect(result.ingredients[0]).toEqual({
        name: "flour",
        quantity: { type: "fixed", value: { type: "decimal", value: 100 } },
        unit: "g",
        optional: false,
        hidden: false,
        preparation: undefined,
        isRecipe: false,
      });
      expect(result.ingredients[1]).toEqual({
        name: "flour",
        quantity: { type: "fixed", value: { type: "decimal", value: 50 } },
        unit: "g",
        optional: false,
        hidden: false,
        preparation: undefined,
        isRecipe: false,
      });
    });

    it("should combine explicitly referenced ingredients case-insensitively", () => {
      const recipe = `
        Add @Sugar{100%g}.
        Then add some more @&sugar{50%g}.
      `;
      const result = new Recipe(recipe);
      expect(result.ingredients).toHaveLength(1);
      expect(result.ingredients[0]).toEqual({
        name: "Sugar", // Note: original casing is preserved
        quantity: { type: "fixed", value: { type: "decimal", value: 150 } },
        unit: "g",
        optional: false,
        hidden: false,
        preparation: undefined,
        isRecipe: false,
      });
    });

    it("should add quantities and convert to largest unit", () => {
      const recipe = `
        Add @sugar{500%g}.
        Then add some more @&sugar{1%kg}.
      `;
      const result = new Recipe(recipe);
      expect(result.ingredients).toHaveLength(1);
      expect(result.ingredients[0]).toEqual({
        name: "sugar",
        quantity: { type: "fixed", value: { type: "decimal", value: 1.5 } },
        unit: "kg",
        optional: false,
        hidden: false,
        preparation: undefined,
        isRecipe: false,
      });
    });

    it("should add quantities and convert to metric", () => {
      const recipe = `
        Add @butter{1%lb}.
        Then add some more @&butter{250%g}.
      `;
      const result = new Recipe(recipe);
      expect(result.ingredients).toHaveLength(1);
      const butter = result.ingredients[0]!;
      expect(butter.name).toBe("butter");
      expect(butter.unit).toBe("kg"); // largest metric mass unit
      expect(butter.quantity).toEqual({
        type: "fixed",
        value: { type: "decimal", value: 0.7 },
      });
      // TODO: 700g would be more elegant
    });

    it("should throw an error if referenced ingredient does not exist", () => {
      const recipe = `Add @&flour{100%g}.`;
      expect(() => new Recipe(recipe)).toThrow(
        /Referenced ingredient "flour" not found/,
      );
    });

    it("adds a referenced ingredient as a new ingredient when units are incompatible", () => {
      const recipe = `
        Add @water{1%l}.
        Then add some more @&water{1%kg}.
      `;
      const result = new Recipe(recipe);
      expect(result.ingredients).toHaveLength(2);
      expect(result.ingredients[0]).toEqual({
        name: "water",
        quantity: { type: "fixed", value: { type: "decimal", value: 1 } },
        unit: "l",
        hidden: false,
        isRecipe: false,
        optional: false,
        preparation: undefined,
      });
      expect(result.ingredients[1]).toEqual({
        name: "water",
        quantity: { type: "fixed", value: { type: "decimal", value: 1 } },
        unit: "kg",
        hidden: false,
        isRecipe: false,
        optional: false,
        preparation: undefined,
      });
    });
  });

  describe("cookware parsing", () => {
    it("extracts cookware correctly", () => {
      const result = new Recipe(simpleRecipe);
      expect(result.cookware.length).toBe(2);
      expect(result.cookware[0]).toEqual({
        hidden: false,
        name: "bowl",
        quantity: { type: "fixed", value: { type: "decimal", value: 1 } },
        optional: false,
      });
      expect(result.cookware[1]).toEqual({
        hidden: false,
        name: "pan",
        optional: false,
      });
    });

    it("should correctly handle modifiers for cookware", () => {
      const recipe = `
        Use an #oven or a #?pan to cook. Clean the #&oven after use. Don't bother using a #-stove
      `;
      const result = new Recipe(recipe);
      expect(result.cookware).toHaveLength(3);
      expect(result.cookware[0]).toEqual({
        name: "oven",
        optional: false,
        hidden: false,
      });
      expect(result.cookware[1]).toEqual({
        name: "pan",
        optional: true,
        hidden: false,
      });
      expect(result.cookware[2]).toEqual({
        name: "stove",
        optional: false,
        hidden: true,
      });
    });
  });

  it("extracts timers correctly", () => {
    const result = new Recipe(simpleRecipe);
    expect(result.timers.length).toBe(1);
    expect(result.timers[0]).toEqual({
      name: undefined,
      duration: { type: "fixed", value: { type: "decimal", value: 15 } },
      unit: "minutes",
    }); // Note: timer name may be empty based on regex
  });

  it("extracts steps correctly", () => {
    const result = new Recipe(simpleRecipe);
    expect(result.sections).toMatchSnapshot();
  });

  it("throws error for missing timer unit", () => {
    const badInput = "Cook for ~{15}";
    expect(() => new Recipe(badInput)).toThrow(/Timer missing units/);
  });

  it("parses notes correctly", () => {
    const recipeWithNotes = `
> This is a note at the beginning.

Add @flour{100%g}
> and this.

> This is a note in the middle
 which continues on the next line.

Another step.

> Another note
> on multiple lines starting with >
> for readability

> A final note.
`;
    const result = new Recipe(recipeWithNotes);
    expect(result.sections).toMatchSnapshot();
  });

  it("parses complex recipes correctly", () => {
    const result = new Recipe(complexRecipe);
    expect(result).toMatchSnapshot();
  });
});
