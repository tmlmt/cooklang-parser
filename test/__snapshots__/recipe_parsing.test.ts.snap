// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse function > parses complex recipes correctly 1`] = `
Recipe {
  "cookware": [
    {
      "hidden": false,
      "name": "Dutch oven",
      "optional": false,
    },
    {
      "hidden": false,
      "name": "small bowl",
      "optional": false,
    },
    {
      "hidden": false,
      "name": "13×9-inch baking dish",
      "optional": false,
    },
    {
      "hidden": false,
      "name": "rubber spatula",
      "optional": false,
    },
    {
      "hidden": false,
      "name": "icing spatula",
      "optional": true,
    },
  ],
  "ingredients": [
    {
      "hidden": false,
      "isRecipe": false,
      "name": "lasagne noodles",
      "optional": false,
      "preparation": undefined,
      "quantity": 9,
      "unit": undefined,
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "bulk Italian sausage",
      "optional": false,
      "preparation": undefined,
      "quantity": 1.2,
      "unit": "lb",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "ground beef",
      "optional": false,
      "preparation": undefined,
      "quantity": 0.75,
      "unit": "lb",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "onion",
      "optional": false,
      "preparation": "medium, diced",
      "quantity": 1,
      "unit": undefined,
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "garlic",
      "optional": false,
      "preparation": undefined,
      "quantity": 3,
      "unit": "cloves",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "crushed tomatoes",
      "optional": false,
      "preparation": undefined,
      "quantity": 2,
      "unit": "cans",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "tomato paste",
      "optional": false,
      "preparation": undefined,
      "quantity": 2,
      "unit": "cans",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "water",
      "optional": false,
      "preparation": undefined,
      "quantity": 0.6666666666666666,
      "unit": "cup",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "sugar",
      "optional": false,
      "preparation": undefined,
      "quantity": 2,
      "unit": "tbsp",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "fresh parsley",
      "optional": false,
      "preparation": "minced, divided",
      "quantity": 0.1,
      "unit": "l",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "dried basil",
      "optional": false,
      "preparation": undefined,
      "quantity": 2,
      "unit": "tsp",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "fennel seed",
      "optional": false,
      "preparation": undefined,
      "quantity": 0.75,
      "unit": "tsp",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "salt",
      "optional": false,
      "preparation": undefined,
      "quantity": 0.75,
      "unit": "tsp",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "pepper",
      "optional": false,
      "preparation": "coarsely ground",
      "quantity": 0.25,
      "unit": "tsp",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "egg",
      "optional": false,
      "preparation": "large, lightly beaten",
      "quantity": 1,
      "unit": undefined,
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "ricotta cheese",
      "optional": false,
      "preparation": undefined,
      "quantity": 1,
      "unit": "carton",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "mozzarella",
      "optional": false,
      "preparation": "shredded",
      "quantity": 4,
      "unit": "cups",
    },
    {
      "hidden": false,
      "isRecipe": false,
      "name": "Parmesan",
      "optional": false,
      "preparation": "grated",
      "quantity": 0.75,
      "unit": "cup",
    },
  ],
  "metadata": {
    "author": "Pam Thompson",
    "cook time": "50m",
    "prep time": "1h",
    "servings": "12, 12 slices",
    "source": "https://www.tasteofhome.com/recipes/best-lasagna/",
    "tags": [
      "pasta",
      "casserole",
      "classic",
    ],
    "time": "1h50m",
    "title": "Best Lasagna",
  },
  "sections": [
    Section {
      "content": [
        {
          "items": [
            {
              "type": "text",
              "value": "Bring a large pot of salted water to a boil. Cook the ",
            },
            {
              "type": "ingredient",
              "value": 0,
            },
            {
              "type": "text",
              "value": " according to the package directions. When complete, drain the noodles, and set aside.",
            },
          ],
        },
        {
          "items": [
            {
              "type": "text",
              "value": "Meanwhile, heat a ",
            },
            {
              "type": "cookware",
              "value": 0,
            },
            {
              "type": "text",
              "value": " over medium heat. Add the ",
            },
            {
              "type": "ingredient",
              "value": 1,
            },
            {
              "type": "text",
              "value": ", ",
            },
            {
              "type": "ingredient",
              "value": 2,
            },
            {
              "type": "text",
              "value": " and diced ",
            },
            {
              "type": "ingredient",
              "value": 3,
            },
            {
              "type": "text",
              "value": ". Cook for ",
            },
            {
              "type": "timer",
              "value": 0,
            },
            {
              "type": "text",
              "value": " or until the meat is no longer pink, breaking up the meat with a large spoon as it cooks to create gorgeous crumbles. Add the ",
            },
            {
              "type": "ingredient",
              "value": 4,
            },
            {
              "type": "text",
              "value": ", and cook for ",
            },
            {
              "type": "timer",
              "value": 1,
            },
            {
              "type": "text",
              "value": ", until the mixture is fragrant.",
            },
          ],
        },
        {
          "note": "Editor’s Tip: To ensure the noodles will not stick to each other, stir the water like a whirlpool when you add the pasta sheets.",
        },
      ],
      "name": "Brown the beef",
    },
    Section {
      "content": [
        {
          "items": [
            {
              "type": "text",
              "value": "Drain the mixture in the Dutch oven. Then, add the remaining sauce ingredients: ",
            },
            {
              "type": "ingredient",
              "value": 5,
            },
            {
              "type": "text",
              "value": ", ",
            },
            {
              "type": "ingredient",
              "value": 6,
            },
            {
              "type": "text",
              "value": ", ",
            },
            {
              "type": "ingredient",
              "value": 7,
            },
            {
              "type": "text",
              "value": ", ",
            },
            {
              "type": "ingredient",
              "value": 8,
            },
            {
              "type": "text",
              "value": ", ",
            },
            {
              "type": "ingredient",
              "value": 9,
            },
            {
              "type": "text",
              "value": ", ",
            },
            {
              "type": "ingredient",
              "value": 10,
            },
            {
              "type": "text",
              "value": ", ",
            },
            {
              "type": "ingredient",
              "value": 11,
            },
            {
              "type": "text",
              "value": ", ",
            },
            {
              "type": "ingredient",
              "value": 12,
            },
            {
              "type": "text",
              "value": " and ",
            },
            {
              "type": "ingredient",
              "value": 13,
            },
            {
              "type": "text",
              "value": ". ",
            },
          ],
        },
        {
          "items": [
            {
              "type": "text",
              "value": "Bring the mixture to a boil. Reduce the heat to a simmer. Cook, uncovered, for ",
            },
            {
              "type": "timer",
              "value": 2,
            },
            {
              "type": "text",
              "value": ", stirring occasionally.",
            },
          ],
        },
      ],
      "name": "Make the meat sauce",
    },
    Section {
      "content": [
        {
          "items": [
            {
              "type": "text",
              "value": "In a ",
            },
            {
              "type": "cookware",
              "value": 1,
            },
            {
              "type": "text",
              "value": ", whisk together the ",
            },
            {
              "type": "ingredient",
              "value": 14,
            },
            {
              "type": "text",
              "value": ", ",
            },
            {
              "type": "ingredient",
              "value": 15,
            },
            {
              "type": "text",
              "value": ", and remaining ",
            },
            {
              "type": "ingredient",
              "value": 9,
            },
            {
              "type": "text",
              "value": " and ",
            },
            {
              "type": "ingredient",
              "value": 12,
            },
            {
              "type": "text",
              "value": ". ",
            },
          ],
        },
        {
          "items": [
            {
              "type": "text",
              "value": "If you haven’t already, grate the ",
            },
            {
              "type": "ingredient",
              "value": 16,
            },
            {
              "type": "text",
              "value": " and ",
            },
            {
              "type": "ingredient",
              "value": 17,
            },
            {
              "type": "text",
              "value": " cheeses.",
            },
          ],
        },
        {
          "note": "Editor’s Tip: For the freshest, creamiest taste, grate your own cheese at home.",
        },
      ],
      "name": "Prepare the cheese filling",
    },
    Section {
      "content": [
        {
          "items": [
            {
              "type": "text",
              "value": "Preheat the oven to 375°. ",
            },
          ],
        },
        {
          "items": [
            {
              "type": "text",
              "value": "Spoon 2 cups meat sauce into an ungreased ",
            },
            {
              "type": "cookware",
              "value": 2,
            },
            {
              "type": "text",
              "value": ", spreading it out to cover the entire dish. Layer three noodles onto the sauce. Add one-third of the ricotta mixture, spreading it over the noodles with a ",
            },
            {
              "type": "cookware",
              "value": 3,
            },
            {
              "type": "text",
              "value": " or an ",
            },
            {
              "type": "cookware",
              "value": 4,
            },
            {
              "type": "text",
              "value": ". Sprinkle with 1 cup mozzarella cheese and 2 tablespoons Parmesan cheese.",
            },
          ],
        },
        {
          "items": [
            {
              "type": "text",
              "value": "Repeat the layers twice, finishing with remaining meat sauce and cheeses. The dish will be very full by the time you get to the top!",
            },
          ],
        },
      ],
      "name": "Layer the lasagna ingredients",
    },
    Section {
      "content": [
        {
          "items": [
            {
              "type": "text",
              "value": "Cover the baking dish, and bake for ",
            },
            {
              "type": "timer",
              "value": 3,
            },
            {
              "type": "text",
              "value": ". ",
            },
          ],
        },
        {
          "items": [
            {
              "type": "text",
              "value": "Remove the cover, and cook for an additional ",
            },
            {
              "type": "timer",
              "value": 3,
            },
            {
              "type": "text",
              "value": ", until the sauce is bubbling around the edges and the cheese is fully melted. ",
            },
          ],
        },
        {
          "items": [
            {
              "type": "text",
              "value": "Let the lasagna stand for ",
            },
            {
              "type": "timer",
              "value": 4,
            },
            {
              "type": "text",
              "value": " before slicing and serving.",
            },
          ],
        },
      ],
      "name": "Bake the lasagna",
    },
  ],
  "servings": 12,
  "timers": [
    {
      "duration": 10,
      "name": undefined,
      "unit": "minutes",
    },
    {
      "duration": 1,
      "name": undefined,
      "unit": "min",
    },
    {
      "duration": 30,
      "name": undefined,
      "unit": "min",
    },
    {
      "duration": 25,
      "name": undefined,
      "unit": "min",
    },
    {
      "duration": 15,
      "name": undefined,
      "unit": "min",
    },
  ],
}
`;
